File: java/L700/LeetCode0721.java
package L700;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LeetCode0721 {
    private class UnionFind {
        private int[] ids, ranks;

        public UnionFind(int N) { 
            ids = new int[N];
            ranks = new int[N];
            for (int i = 0;i < N;++i) {
                ids[i] = i;
            }
        }

        public int find(int id) {
            while (ids[id] != id) {
                id = ids[id];
            }
            return id;
        }

        public void union(int p, int q) {
            int pid = find(p);
            int qid = find(q);
            if (pid != qid) {
                if (ranks[pid] <= ranks[qid]) {
                    ids[pid] = qid;
                } else {
                    ids[qid] = pid;
                }
                if (ranks[pid] == ranks[qid]) {
                    ranks[qid] += 1;
                }
            }
        }
    }
    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        Map<String, Integer> emailIndex = new HashMap<>();
        Map<String,String> emailName = new HashMap<>();
        int emailCount = 0;
        for (List<String> account : accounts) {
            String name = account.get(0);
            int sz = account.size();
            for (int i = 1;i < sz;++i) {
                String email = account.get(i);
                if (!emailIndex.containsKey(email)) {
                    emailIndex.put(email, emailCount);
                    ++emailCount;
                    emailName.put(email, name);
                }
            }
        }
        UnionFind uf = new UnionFind(emailCount);
        for (List<String> account : accounts) {
            String firstEmail = account.get(1);
            int sz = account.size();
            for (int i = 2;i < sz;++i) {
                uf.union(emailIndex.get(firstEmail), emailIndex.get(account.get(i)));
            }
        }
        Map<Integer, List<String>> indexEmail = new HashMap<>();
        for (Map.Entry<String, Integer> entry : emailIndex.entrySet()) {
            int id = uf.find(entry.getValue());
            List<String> cur = indexEmail.getOrDefault(id, new ArrayList<>());
            cur.add(entry.getKey());
            indexEmail.put(id, cur);
        }
        List<List<String>> res = new ArrayList<>();
        for (Map.Entry<Integer,List<String>> entry: indexEmail.entrySet()) {
            List<String> emails = entry.getValue();
            Collections.sort(emails);
            List<String> cur = new ArrayList<>();
            cur.add(emailName.get(emails.get(0)));
            for (String email : emails) {
                cur.add(email);
            }
            res.add(cur);
        }
        return res;
    }
}

File: cpp/L700/LeetCode0721.cc
#include <algorithm>
#include <string_view>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using std::vector;
#include <string>
using std::string;

struct UnionFind {
    vector<int> ids, weights;

    UnionFind(int N): ids(N, 0), weights(N, 1) {
        for (int i = 0;i < N;++i) {
            ids[i] = i;
        }
    }

    int toFind(int id) {
        while (ids[id] != id) {
            id = ids[id];
        }
        return id;
    }

    void toUnion(int p, int q) {
        int pid = toFind(p);
        int qid = toFind(q);
        if (pid == qid) {
            return;
        }
        if (weights[pid] >= weights[qid]) {
            ids[qid] = pid;
            weights[pid] += weights[qid];
        } else {
            ids[pid] = qid;
            weights[qid] += weights[pid];
        }
    }
};

class Solution {
    bool existEmail(vector<string> const &v1, vector<string> const &v2) {
        for (int i = 1;i < v1.size();++i) {
            for (int j = 1;j < v2.size();++j) {
                if (v1[i] == v2[j]) {
                    return true;
                }
            }
        }
        return false;
    }
public:
    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        int N = accounts.size();
        if (N <= 0) {
            return {};
        }
        UnionFind uf(N);
        for (int i = 0;i < N;++i) {
            for (int j = i + 1;j < N;++j) {
                if (existEmail(accounts[i], accounts[j])) {
                    uf.toUnion(i, j);
                }
            }
        }
        std::unordered_map<int, std::unordered_set<std::string_view>> mp;
        for (int i = 0;i < N;++i) {
            int id = uf.toFind(i);
            auto iter = mp.find(id);
            if (iter == mp.end()) {
                std::unordered_set<std::string_view> st;
                for (int j = 1;j < accounts[id].size();++j) {
                    st.emplace(accounts[id][j]);
                }
                mp[id] = st;
            }
            iter = mp.find(id);
            for (int j = 1;j < accounts[i].size();++j) {
                iter->second.emplace(accounts[i][j]);
            }
        }
        vector<vector<string>> res;
        for (auto const &[id, st] : mp) {
            res.push_back(vector<string>());
            res.back().push_back(accounts[id][0]);
            for (std::string_view sv : st) {
                res.back().emplace_back(string(sv.begin(), sv.end()));
            }
            std::sort(res.back().begin() + 1, res.back().end());
        }
        return res;
    }
};